pkg/controller/operconfig/migration.go

```diff
@@ -97,3 +120,325 @@ func convertEgressFirewallToEgressNetworkPolicy(ctx context.Context, client cnoc
+		}
+
+		if hostSubnetHasEgressIpConfigManual {
+			log.Printf("Manual configuration of SDN egressIP detected and is unsupported for migration; OVN egressIPs will be generated but will not maintain individual node assignments from SDN hostsubnets")
+		}
+	}
+
+	if !hostSubnetFound {
+		log.Printf("did not find a hostsubnet object with egressIP configured, quitting process early")
+		return nil
+	} else {
+		log.Printf("found hostsubnet object with egressIP configured, continuing...")
+	}
+
+	// 3. iterate through netnamespaces
+	//    - any with egressIP configured will cause an egressIP ovn resource to be created via k8s api
+	//    - a corresponding namespace label will be added to match the egressIP resource's namespace selector field
+	for i, nns := range netNamespaceList {
+		if netNamespaceHasEgressIpConfig(*nns) {
+			// config detected, translating to ovn
+			// - delete cloudprivateipconfig if it exists
+			// - create egressip custom resource for ovn
+			if err := deleteCloudPrivateIpConfigs(ctx, client, nns.Object["egressIPs"].([]interface{})); err != nil {
+				return err
+			}
+
+			egressIpName := fmt.Sprint("egressip-", i)
+			egressIP := unstructuredEgressIpObject(egressIpName, nns.Object["egressIPs"].([]interface{}), nns.Object["netname"])
+			if err := apply.ApplyObject(ctx, client, egressIP, ""); err != nil {
+				return err
+			}
+
+			// now remove egressIP config from netnamespace
+			nns.Object["egressIPs"] = nil
+			if _, err := client.Default().Dynamic().Resource(gvrNetnamespace).Update(ctx, nns, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil // success
+}
+
+func convertOvnEgressIpToSdnEgressIp(ctx context.Context, client cnoclient.Client) error {
+	egressIpRollbackReady, err := sdnEgressIpResourcesReady(ctx, client)
+	if !egressIpRollbackReady {
+		return nil // wait for all SDN egressIP resources to be created before rollback
+	} else if err != nil {
+		return err
+	}
+
+	// 1. query for egressips
+	egressIpList, err := ListAllOfSpecifiedType(gvrEgressIp, ctx, client)
+	if err != nil {
+		return err
+	}
+
+	// 2. query for nodes
+	nodeList, err := ListAllNodes(ctx, client)
+	if err != nil {
+		return err
+	}
+
+	// 3. iterate through egressIP objects and generate netnamespaces
+	//    - value of `namespaceSelector` will determine netnamespace's name
+	//    - egressIP list will be copied to egressIPs field for netnamespace
+	for _, eip := range egressIpList {
+		// translate egressIPs to SDN netnamespaces
+		// - delete cloudprivateipconfig if it exists
+		// - create egressip to netnamespace resources for sdn
+
+		egressIps := eip.Object["spec"].(map[string]interface{})["egressIPs"].([]interface{})
+
+		if err := deleteCloudPrivateIpConfigs(ctx, client, egressIps); err != nil {
+			return err
+		}
+
+		eipNamespace := eip.Object["spec"].(map[string]interface{})["namespaceSelector"].(map[string]interface{})["matchLabels"].(map[string]interface{})["kubernetes.io/metadata.name"]
```

Comment:
Addressing `matchLabels` is an unsafe access as it is not required.

I guess we are assuming that these are only the egress ips that we previously migrated, but the code could be more robust anyway.

Overall I would have liked to see more use of [unstructure](https://pkg.go.dev/k8s.io/apimachinery@v0.25.0/pkg/apis/meta/v1/unstructured) to access unstructured objects, then we would not have to review or guess what and when is something required and must be there and which might not, and we get conversions handled for us for the most common data types.

Reply:
> I guess we are assuming that these are only the egress ips that we previously migrated

Yes this is correct, so it is not an unsafe access if the user is rolling back according to what we allow in documentation.

> Overall I would have liked to see more use of [unstructure](https://pkg.go.dev/k8s.io/apimachinery@v0.25.0/pkg/apis/meta/v1/unstructured) to access unstructured objects

Perhaps this can be addressed later in a separate PR.

Reply:
So this is not only a stilistic type of suggestion, [unstructure](https://pkg.go.dev/k8s.io/apimachinery@v0.25.0/pkg/apis/meta/v1/unstructured) does the correct safety access check on the maps and uses type casting for conversions instead of Sprintf.

An example could be:
```
eipNamespace, found, err := unstructured.NestedString(eip.Object, "spec", "namespaceSelector", "matchLabels", "kubernetes.io/metadata.name")
if !found || err != nil {
    return fmt.Errorf(err, "kubernetes.io/metadata.name not found in egress ip object %s, probable underlying error: %v", eipName, err)
} 
```

in this case though `matchLabels` is defined as a string and not as an object so I am not sure it can be accessed as a `map[string]interface{}` but if your code works then this example should work as well.

Reply:
From the egressIP yaml definition:
```
matchLabels:
  additionalProperties:
    type: string
  description: matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels map is equivalent to an element of matchExpressions, whose key field is "key", the operator is "In", and the values array contains only "value". The requirements are ANDed.
  type: object
```  

matchLabels is defined as a map object, not as a string.

Like you said matchLabels is optional, so I went ahead and replaced this instance with your example since it's a fair point. Other instances either have checks on optional fields or only check required fields, and they are significantly less verbose, so I left those unchanged.